### PR TITLE
Upgrade to pydantic v2 

### PR DESCRIPTION
**Pull request by oscarbc96**
Depends on: https://github.com/Skyscanner/pycfmodel/pull/141

### CODE REVIEW

File: tests/rules/test_WildcardResourceRule.py [hunk 1]
```diff
@@ -661,16 +683,38 @@ def test_multiple_resources_with_wildcard_resources_are_detected(user_and_policy
...
+                    "dynamodb:BatchWrite*",
+                    "dynamodb:DescribeTable",
+                    "dynamodb:Scan",
                     "dynamodb:PutItem",
+                    "dynamodb:Get*",
+                },
+                resource_ids={"RolePolicy"},
+                resource_types={"AWS::IAM::Policy"},
+            ),
+            Failure(
```
Comment (oscarbc96) [inline tests/rules/test_WildcardResourceRule.py:701]: Two new failures are added. Review if they are correct

File: tests/rules/test_S3LifecycleConfigurationRule.py [hunk 2]
```diff
@@ -17,7 +17,6 @@ def bad_template_no_configuration():
     "template_path",
     [
         "rules/S3LifecycleConfiguration/good_template.yaml",
-        "rules/S3LifecycleConfiguration/allowed_template_malformed_lifecycle_rules.yaml",
```
Comment (oscarbc96) [inline tests/rules/test_S3LifecycleConfigurationRule.py:20]: If it's a malmormed policy, pycfmodel must not parse it. This test is wrong.

File: tests/rules/test_WildcardResourceRule.py (hunk 1 above)
Comment (ignaciobolonio) [reply tests/rules/test_WildcardResourceRule.py:701]: They are, they come from https://github.com/Skyscanner/pycfmodel/pull/133 which is not included in the latest release yet

**Review by ignaciobolonio — DISMISSED** [review]
Looking great 🚀

**Comment by ignaciobolonio** [comment]
will continue in #289 as for some reason tests are not being run as we expected

**Pull request merged**